Extract, unit test, and improve getLabelFromStackTrace

srmagura opened the pull request:
<!--
Thanks for your interest in the project. I appreciate bugs filed and PRs submitted!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: http://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**:

There were a few of cases where extracting the label from the stack trace did not work correctly.
- This one is still an issue (Safari): #2255.
- Closes #2226. 
- Closes #2134.

**Why**:

- The stack trace parsing code is complex but was lacking tests.
- There were multiple reports of labels being different between SSR and client-side rendering during development.

**How**:

- Create function `getLabelFromStackTrace` to encapsulate the regex logic.
- Write unit tests for it.
- Playgrounds:
    - Recreate the CRA playground which was on react-scripts v1.
    - Add Next.js playground for testing SSR.
    - Remove [broken](https://pastebin.com/P4a3GhaW) Razzle playground. If it is needed in the future, we can just recreate it.

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->
<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [x] Documentation
- [x] Tests
- [x] Code complete
- [x] Changeset

### Before changing the implementation of `getLabelByStackTrace`

![tests-before](https://user-images.githubusercontent.com/801549/139869047-b3e78725-bede-4340-9428-da2370d03079.png)

### After changing the implementation

All tests pass.

### Safari is weird

_This section is now outdated._

It looks like Safari likes to inline function calls, resulting in that function being omitted from the stack trace. In both of the following cases, there will be a "class name mismatch" warning if SSR is used.

- `typical function component - runtime=classic, Safari`: component name does not appear in stack trace. The label should be `undefined` in this case.
    - I'm not worried about this because it only affects the old JSX runtime. 
- `function component within object, Safari`: component name is omitted from stack trace. The label should be the parent component's name in this case.
    - Andarist, do you have any idea how common this pattern is in practice? `const test = { MyComp() { return <div css={{}} /> } }`. If this is a corner case, we may not have to worry about it.

If either of these are going to be a problem, we could introduce a feature flag that disables label extraction. Label extraction would still be on by default.

Andarist commented:
Amazing work so far, thank you ❤️ 

I would also take a look at this PR and the included comments: https://github.com/emotion-js/emotion/pull/2136/files . It's fine to continue with our current strategy - just mentioning this for completeness, might give you some new ideas/test cases.

Please ping me when this will be ready for review. As far as I can see you are still working on this one.

Andarist commented on packages/react/src/emotion-element.js:
```diff
@@ -11,6 +11,26 @@ import { serializeStyles } from '@emotion/serialize'
+
+  // console.log(stackTrace)
+
+  // Chrome — TODO: why is it like this?
```
`React.forwardRef` creates an object with a `render` property that is later on called by React. If we mimic this with:
```js
onst test = { render: function MyComp() { console.log(new Error().stack) } }
test.render()
```
we'll receive such a stack trace:
```
Error
    at Object.MyComp [as render] (<anonymous>:1:56)
    at <anonymous>:2:6
```

This was fixed in this PR: https://github.com/emotion-js/emotion/pull/1973

As to the `(?:Object\.|Module\.|)` part. If we use this:
```js
const test = { MyComp() { console.log(new Error().stack) } }
test.MyComp()
```
then we end up with:
```
Error
    at Object.MyComp (<anonymous>:1:39)
    at <anonymous>:2:6
```

There is also a chance that a component gets exported from an actual module and this is when the Module part kicks in. The Module part was originally added in response to this issue: https://github.com/emotion-js/emotion/issues/1947

We can also find such a test case from the old issues: https://github.com/emotion-js/emotion/pull/1714#discussion_r365518850

srmagura commented:
Thank you!!! The links were very helpful. 😁

I will let you know when the work is done.

srmagura commented:
@Andarist   What are your thoughts on my "Safari is weird" section in the original post?

Andarist commented on packages/react/__tests__/get-label-from-stack-trace.js:
````diff
@@ -0,0 +1,457 @@
+ * E.g.
+ *
+ * ```
+ * const test = {
````
it would be great to have such an example before each describe block - some of them are missing it and it makes it slightly harder to analyze.

Andarist commented on packages/react/__tests__/get-label-from-stack-trace.js:
```diff
@@ -0,0 +1,457 @@
+  })
+
+  test('Safari', () => {
+    // Strangely, the component name does not appear in the stacktrace
```
oh my - this is indeed weird, what's the difference between the classic and automatic runtimes here? o.O

srmagura replied:
Yeah I have no idea LOL! I will retest this one tomorrow.

Andarist commented:
> typical function component - runtime=classic, Safari: component name does not appear in stack trace. The label should be undefined in this case.

This is bizarre 😢 Especially that I don't see what's the difference between both runtimes here - their stack traces have the same structure in all browsers, only the jsx factory name is different.

> Andarist, do you have any idea how common this pattern is in practice? const test = { MyComp() { return <div css={{}} /> } }. If this is a corner case, we may not have to worry about it.

AFAIK this is not common so I think it's OK to not care about it right now. If we could distinguish this situation in other stack traces reliably to return undefined there then that would be great. But if that's too problematic - then it shouldn't block the work done here.

srmagura commented:
@Andarist This is ready for review! 😁

- Safari is SO weird.... see my notes about it in `__tests__/get-label-from-stack-trace.js` and the documentation I edited. You were right about it having nothing to do with whether the new or old JSX transform is used. I talked to my coworker about the function names being omitted from the stack trace and he thinks this a straight up bug in the Safari JS engine (though I doubt the team would agree with this). I suppose I could report it on the WebKit Bugzilla but that would probably be a waste of time.
- I looked into why the netlify checks are failing, and, big surprise, it is failing to install sharp. Can this be ignored until I get around to upgrading the website?
- `yarn lint:check` seems to just run forever on both my Windows PC and my Mac. It works in GitHub Actions so I guess it's not really a problem.
    ![image](https://user-images.githubusercontent.com/801549/140100637-8099d172-a2de-46bd-8473-14a50fa459c8.png)
- I think we can close #2136.

After this is merged, I need to focus on my actual job for the next 2 weeks. But I'll be back to work on the website upgrade!

mitchellhamilton commented:
re the Safari thing, it's not a bug, it's because JavaScriptCore is the only engine to implement Proper Tail Calls (https://webkit.org/blog/6240/ecmascript-6-proper-tail-calls-in-webkit/)

srmagura commented:
OMG that makes so much sense 🤯🤯🤯. I'll update my comments & documentation to reflect this.

Though, AFAIK we have no way to prevent tail call optimization so label extraction may still fail in JS Core.

srmagura commented:
I updated the documentation and comments regarding Safari stack traces.

In the last day, there have been multiple people asking about problems related to label extraction:

- #2255: Asking if there is a fix for "class name mismatch" in Safari. (There is no way to fix this because of Proper Tail Calls.)
- #2303: Asking for the ability to disable label extraction to improve performance and make performance in dev more indicative of performance in prod. (It's _possible_ that my PR will improve the performance. A quick & dirty test showed `getLabelFromStackTrace` taking approximately 1 µs.)

Should we add a feature flag that disables label extraction? Label extraction would be ON by default.

Andarist commented on playgrounds/nextjs/pages/index.js:
```diff
@@ -0,0 +1,31 @@
+/** @jsxImportSource @emotion/react */
```
you have configured `importSource` in the Babel config here, so the explicit pragma shouldn't be needed

Andarist commented on docs/labels.mdx:
````diff
@@ -30,3 +30,19 @@ render(
 )
 ```
+
+## SSR & Safari
````
I wonder if there is any way to opt-out of this optimization in our dev build. I don't see how this would be possible as it is applied to a function (a component) that is not under our control.

The optimization kicks in when the returned element from a component is the CSS prop element. Like here:
```jsx
function Home() {
  return <div css={{ color: 'orchid' }}>Some colored text</div>
}
```

It is enough to not return this element directly to prevent this optimization from kicking in:
```jsx
function Home() {
  const el = <div css={{ color: 'orchid' }}>Some colored text</div>
  return el
}
```
but as mentioned - the `Home` function in this example is out of our control so I'm not sure how we possibly could prevent it from happening 🤔 

Maybe we should reach for [`suppressHydrationWarning`](https://reactjs.org/docs/dom-elements.html#suppresshydrationwarning) in this situation?

This creating a confusing warning for the user in one of the browsers is IMHO really bad - to the point that I don't think that having this logic with the warning is worth it.

@mitchellhamilton any thoughts on the subject? what do you think about dropping this runtime extraction or using `suppressHydrationWarning` in Safari?

Andarist replied:
We could also try to minimize the need for the suppression prop by consulting the DOM, some labels could be extracted from the SSRed HTML

Andarist replied:
Well, the suppression flag is a dumb idea as it would most likely hide other, legit, hydration mismatches that could happen there

srmagura replied:
> We could also try to minimize the need for the suppression prop by consulting the DOM, some labels could be extracted from the SSRed HTML

Very interesting idea. It seems like it would really hard to find the correct DOM node, but if we could do that the label extraction would be simple.

The only other options I can think of are:
- A global flag to either opt-in or opt-out of label extraction.
- Remove label extraction period. I don't like this since the labels are very useful.

But I defer to you and Mitchell.

mitchellhamilton replied:
> suppressHydrationWarning

yeah, i'm not really into doing that

> We could also try to minimize the need for the suppression prop by consulting the DOM, some labels could be extracted from the SSRed HTML

How are you thinking this could work?

> what do you think about dropping this runtime extraction

tbh, I've been kinda wanting to do this because I've got a project where I render a _lot_ of elements and the label extraction slows it down to the point where it's faster for me to do a production build and use that than using the dev version(with the label extraction, it takes ~1.2 minutes vs ~10 seconds with the label extraction removed from the package but still in dev to hydrate the tree and basically all that cost is the regex). If we do want to keep it, we should really improve the performance. (Here are traces with and without the label extraction if you're curious, you'll need to unzip them before loading them into Chrome devtools - [with-label-extraction.json.zip](https://github.com/emotion-js/emotion/files/7535028/with-label-extraction.json.zip), [without-label-extraction.json.zip](https://github.com/emotion-js/emotion/files/7535030/without-label-extraction.json.zip))

Andarist replied:
> How are you thinking this could work?

The problem with this solution is that it won't work perfectly for:
```js
const style = css`color: hotpink;`

function Comp1() { return <div css={style} /> }
function Comp2() { return <div css={style} /> }
```

It's because this now results in 2 different labels - but in just a single hash. This doesn't sound like a super common situation though.

Since the whole thing in Safari can only affect SSR scenarios we should be able to extract the label from the data attribute on `document.querySelector('[data-emotion*="' + hash + '"]').dataset.emotion`. 

At the moment the regexp parsing happens in the `createEmotionProps`, while serializing styles happen in the Emotion element - so this thing I'm talking about would have to be added to the latter. 

Unsure if this is a good idea - just an idea I've had.

> If we do want to keep it, we should really improve the performance. (Here are traces with and without the label extraction if you're curious, you'll need to unzip them before loading them into Chrome devtools - with-label-extraction.json.zip, without-label-extraction.json.zip)

could you share actual stack traces? it's not totally straightforward to grab them from the profiles - or I don't know how to do that. I see though that this takes a lot of time for u there! 😱 

@srmagura has refactored this extraction quite a bit here and it has potential of being fast, we could also try to optimize this even further. I think it's worth to give this a shot - even to learn how fast we can get with that and how much of an impact this kind of a feature can have on the runtime. Would you be willing to try out a snapshot release in your project?

srmagura replied:
I'm happy to benchmark the new implementation against Mitchell's stack traces, but I think we should first decide if label extraction is being removed. No point in optimizing code if we end up deleting it.

That said, I think the new stack trace parsing should already be pretty fast, with the potential for further optimization.

mitchellhamilton replied:
I tried this branch on my project and it totally removed the regex being visible from the trace so that's great and it's  down to within a couple seconds of the version without the label extraction.

Also, applying the patch below and adding labels to some things practically completely removed the difference between the version with and without label extraction given it's basically never running.

```patch
diff --git a/packages/react/src/emotion-element.js b/packages/react/src/emotion-element.js
index d2f97cd1..df5f689b 100644
--- a/packages/react/src/emotion-element.js
+++ b/packages/react/src/emotion-element.js
@@ -33,7 +33,12 @@ export const createEmotionProps = (type: React.ElementType, props: Object) => {
 
   newProps[typePropName] = type
 
-  if (process.env.NODE_ENV !== 'production') {
+  if (
+    process.env.NODE_ENV !== 'production' &&
+    (typeof props.css !== 'object' ||
+      typeof props.css.name !== 'string' ||
+      props.css.name.indexOf('-') === -1)
+  ) {
     const label = getLabelFromStackTrace(new Error().stack)
     if (label) newProps[labelPropName] = label
   }
```

> we should first decide if label extraction is being removed

I think the problems with it are:

- Performance
- The Safari thing

Given the changes already in this PR, the performance impact is now pretty minimal and given this is in dev, I think a small performance impact is totally fine.
I'm not that concerned about the Safari issue given it's just a warning.
Also, given the patch above + using the Babel plugin, the performance impact doesn't really exist since you'll basically always skip that code and you won't get problems on Safari given the labels are added by the Babel plugin.

So I think I'm okay with leaving it in unless you disagree @Andarist?

mitchellhamilton replied:
An idea for addressing the Safari issue: what if we overwrite `console.error` and don't call the real `console.error` warnings when we know they're fine?

Ideally we would even look at the content of the message, get the client side class name, look in the cache for the content for that, add `label:LabelExtractedFromServerSideClassName;`, hash it and compare the names and only skip it when they match so you'd still get the warning when it's correct

srmagura replied:
I put in your performance improvement @mitchellhamilton! (with an added null check)

I like the `console.error` idea and can try that out on a new branch. Am I right in thinking that I will have to overwrite `console.error` once per Emotion cache? Since there could be multiple caches used on a single page.

mitchellhamilton replied:
> Am I right in thinking that I will have to overwrite console.error once per Emotion cache?

Yes

Andarist replied:
> Also, applying the patch below and adding labels to some things practically completely removed the difference between the version with and without label extraction given it's basically never running.

Could you confirm that the performance improvement was still huge without this patch? From what I read from it - it checks if the value passed to the css prop has been transformed by our Babel plugin, right? Since there is, indeed, no need to get the label from the stack trace in such a case - this is a good optimization for Babel users. I'd like to make sure that the performance improvements prepared by @srmagura are enough for non-Babel users. I'm pretty sure that they are - just double-checking.

> So I think I'm okay with leaving it in unless you disagree @Andarist?

Assuming the perf of this is fine then I'm OK with leaving this as is - it's a useful feature. Somebody has suggested in the past that sometimes mere `new Error().stack` can be a major perf hit but I was never able to experience that. No profile dumps that I've investigated showed that.

> An idea for addressing the Safari issue: what if we overwrite console.error and don't call the real console.error warnings when we know they're fine?

This has crossed my mind but I've concluded that something like this is always risky - although probably would work in a vast majority of cases. I would have to recheck but I think that, for instance, React DevTools might also overwrite those. I don't see how we could, for example, clash with their logic but if we could stay away from that - it would be IMHO preferable. 

What do you think about the querySelector approach that would attempt to minimize the mismatch? With proper caching and stuff it feels like it could take us quite far and **without** losing the label - which is tempting. Given that you propose rehashing stuff from within the overridden `console.error` etc I'm not sure if the `console.error` approach would end up being much simpler, implementation-wise. An argument in favor of the `console.error` approach would definitely be that we would never report any "false positives" to users - although with the querySelector approach we potentially could assess if the mismatch can happen (in rare, rare cases) and print a helpful information to the console that one should ignore the React warning and that this can be "fixed" by using a Babel plugin.

Andarist replied:
btw. nice teamwork everybody! 

<img src="https://media.giphy.com/media/xUPGcjQ6dJEjH5uwMw/giphy-downsized-large.gif" />

srmagura replied:
Love the teamwork 😎. 

There are definitely other tools that overwrite `console.error` e.g. error monitoring libraries like Bugsnag. The risk I see is that the "class name mismatch" error could get logged somewhere before we have a chance to intercept it.

The querySelector approach seems simpler but, as you said, it does not handle this case:

```ts
const style = css`color: hotpink;`

function Comp1() { return <div css={style} /> }
function Comp2() { return <div css={style} /> }
```

I'm also not sure how common this is, but it is a valid use case. Maybe you have hotpink text all over your site and want to just import that same `style` object everywhere. We would end up giving all these elements the same label which would be confusing.

I think `console.error` would be my preference.

BTW, could we merge this now and I'll do a separate PR for working around the Safari issue? Since I believe the new `getLabelByStackTrace` is an improvement over what's in `main` despite it not solving the Safari issue.

Andarist replied:
> I'm also not sure how common this is, but it is a valid use case. Maybe you have hotpink text all over your site and want to just import that same style object everywhere. We would end up giving all these elements the same label which would be confusing.

Importing the same style object is very much a common practice - I just assume that in such situations such an object is mixed together with some other styles. Since we "flatten" all the styles into a single class name the resulting hash should be different and thus maybe this would work almost perfectly in practice. Hard to tell though as we don't have any good data on how people are writing their styles in the real world.

mitchellhamilton replied:
> Could you confirm that the performance improvement was still huge without this patch?

yes, it was

> What do you think about the querySelector approach that would attempt to minimize the mismatch?

I don't really see how it would work, even ignoring styles being used with different labels, given that the hash is based on the style with the `label:Blah;` how would we match a style to a style with a label?

> This has crossed my mind but I've concluded that something like this is always risky - although probably would work in a vast majority of cases. I would have to recheck but I think that, for instance, React DevTools might also overwrite those. I don't see how we could, for example, clash with their logic but if we could stay away from that - it would be IMHO preferable.

React DevTools does override console.error but yeah, I don't see how it would clash. Maybe depending on which one overrides first, that could break things?

Andarist replied:
> I don't really see how it would work, even ignoring styles being used with different labels, given that the hash is based on the style with the label:Blah; how would we match a style to a style with a label?

🤦 I forgot that the label is part of the hashed content. 

> React DevTools does override console.error but yeah, I don't see how it would clash. Maybe depending on which one overrides first, that could break things?

Ye, maybe - hard to assess this without actually trying this out

mitchellhamilton commented on packages/react/src/get-label-from-stack-trace.js:
```diff
@@ -0,0 +1,53 @@
+  return undefined
+}
+
+const internalReactFunctionNames = new Set([
```
```suggestion
const internalReactFunctionNames = /* #__PURE__ */ new Set([
```

Do you think minifiers will be able to remove all this code @Andarist? Given they're all just declaring functions, I would imagine it should be fine?

Andarist replied:
IIRC this should be removed if we add the PURE comment here, the passed argument is pure so that part is fine - it's only this constructor call that is "dangerous". We already add the PURE annotation in very similar place, like here:
https://github.com/emotion-js/emotion/blob/2bac69b6c058d007cfc190e94490c0cb225ca40c/packages/cache/src/stylis-plugins.js#L86

And the rest is totally fine as those are just function declarations/expressions so they should be easily removed if they stay unreferenced.

srmagura commented:
Btw, I tested this in IE and the new code didn't cause any errors. Of course, the runtime label extraction does not actually work — `new Error().stack` is actually `undefined`.

srmagura commented:
Removed the "SSR & Safari" section from the docs since we are planning to fix that one way or another. I also updated the initial comment of the PR to not close the issue about class name mismatch in Safari, since that is still an issue.

This should be ready for merge as far as I'm concerned. Thanks.

Andarist commented on .changeset/sweet-hotels-explain.md:
```diff
@@ -0,0 +1,5 @@
+'@emotion/react': patch
+---
+
+Changed the implementation of label extraction from stack traces to one that should yield more consistent results across browsers.
```
```suggestion
Changed the implementation of the runtime label extraction in element using css prop (that only happens in development) from stack traces to one that should yield more consistent results across browsers. This fixes some minor issues with React reporting hydration mismatches that wouldn't happen in production.
```

Andarist commented on packages/react/__tests__/get-label-from-stack-trace.js:
```diff
@@ -0,0 +1,560 @@
+  getLabelFromStackTrace
+} from '../src/get-label-from-stack-trace'
+
+describe('getFunctionNameFromStackTraceLine', () => {
+  test('V8', () => {
+    expect(getFunctionNameFromStackTraceLine('Error')).toBeUndefined()
+
+    expect(
+      getFunctionNameFromStackTraceLine(
+        '    at Object.createEmotionProps (emotion-element-1fb5ab00.browser.esm.js:143)'
+      )
+    ).toBe('createEmotionProps')
+  })
+
+  test('Safari / Firefox', () => {
+    expect(
+      getFunctionNameFromStackTraceLine(
+        'createEmotionProps@http://localhost:3000/static/js/main.chunk.js:844:40'
+      )
+    ).toBe('createEmotionProps')
+  })
+})
```
I probably think that this is a little bit redundant because it's sort of an implementation detail. It makes sense that we have to test the whole thing using the unit test approach because there is no other way to do it. However, I feel like we should always try to focus on the outermost "boundary". There are cases when units makes sense - sometimes it's just nice to isolate some tricky, convoluted, logic and test it in separation. Especially if it is consumed by many other places in the code.

This is just a nit, and a personal preference - so it's fine as is. It just kinda doesn't match the style of testing used in the rest of the codebase.

Andarist commented on packages/react/__tests__/get-label-from-stack-trace.js:
```diff
@@ -0,0 +1,560 @@
+ * All tests after this block are for runtime=automatic, i.e. the new JSX
+ * transform introduced in React 17
+ */
+describe('typical function component - runtime=classic', () => {
```
q: shouldn't this describe block contain a SSR case too?

Andarist commented on packages/react/__tests__/get-label-from-stack-trace.js:
```diff
@@ -0,0 +1,560 @@
+  })
+
+  test('SSR', () => {
+    const stackTrace = `Error
```
huh, it's interesting that the same case has produced different-looking stack traces in Chrome and in node, even though they are both using V8 🤔

srmagura replied:
It is interesting. It is likely because the Chrome stack trace is from CRA while the SSR one is from Next.js and there are differences in build tooling. Let me know if you want me to redo the Chrome one using the Next.js playground.

Andarist commented on packages/react/__tests__/get-label-from-stack-trace.js:
```diff
@@ -0,0 +1,560 @@
+  })
+
+  test('Safari', () => {
+    const stackTrace = `createEmotionProps@http://localhost:3000/static/js/bundle.js:46440:49
```
is this stack trace any different - format-wise - than other stack traces in this block? it looks roughly the same as FireFox one, right?

what I'm getting at - if any of the Safari stack traces are different here we should still test them in a way that they won't result in the `undefined` label. To do that we can simply move the element out of the return position of the function, assign it to a variable and return that variable - this prevents TCO

srmagura replied:
The Safari stack appears identical to the Firefox one apart from the component function call being omitted. So I don't think there's any need to test Safari without TCO. Let me know if you disagree and I will change it.

Andarist commented on packages/react/__tests__/get-label-from-stack-trace.js:
```diff
@@ -0,0 +1,560 @@
+  test('Safari', () => {
+    const stackTrace = `createEmotionProps@http://localhost:3000/static/js/main.chunk.js:844:49
+jsxDEV@http://localhost:3000/static/js/main.chunk.js:1126:247
+MyComponent$9@http://localhost:3000/static/js/main.chunk.js:2274:92
```
this is somewhat interesting, why the TCO has not been applied here? 🤔

srmagura replied:
Yeah it is odd. I can redo this one using the exact code from the comment block if you want. I've noticed that very minor things can affect whether TCO is performed.

Andarist commented on packages/react/__tests__/get-label-from-stack-trace.js:
```diff
@@ -0,0 +1,560 @@
+  })
+
+  test('Safari', () => {
+    const stackTrace = `createEmotionProps@http://localhost:3000/static/js/main.chunk.js:844:49
```
that's a funny case :P since in here the TCOed function is not relevant for us, it still works in the Safari - which is kinda of no surprise

Andarist commented on packages/react/__tests__/get-label-from-stack-trace.js:
```diff
@@ -0,0 +1,560 @@
+    // MyComponent$9 does not appear in the stack trace
+    const stackTrace = `createEmotionProps@http://localhost:3000/static/js/main.chunk.js:866:49
+jsxDEV@http://localhost:3000/static/js/main.chunk.js:1147:247
+App@http://localhost:3000/static/js/main.chunk.js:2290:32
```
this is gonna be hard(er) to handle in the potential console.warn/error override - since in here we extract the label, it's just a wrong one

srmagura replied:
Agreed. I _think_ I will still be able to suppress the warning in this case...

Andarist commented on packages/react/__tests__/get-label-from-stack-trace.js:
```diff
@@ -0,0 +1,560 @@
+  })
+})
+
+test('Chrome: Module.jsx', () => {
```
could you add a note above this block mentioning how one reproduce this situation?

srmagura replied:
I never encountered a stack trace containing `Module.` during all of my testing so I'm not sure how this happens. It's complicated because it could depend on stuff like which version of Webpack is used.

Andarist commented on packages/react/__tests__/get-label-from-stack-trace.js:
````diff
@@ -0,0 +1,560 @@
+ * ```
+ */
+describe('React.forwardRef with named function', () => {
+  // forwardRef only changes the stack trace for SSR
````
interesting - do u know why? I think i've encountered this before but I forgot what the reason it

srmagura replied:
No idea lol, I just reached this conclusion via experimentation.

Andarist commented on packages/react/src/emotion-element.js:
```diff
@@ -37,21 +33,17 @@ export const createEmotionProps = (type: React.ElementType, props: Object) => {
+  // the label hasn't already been computed
+  if (
+    process.env.NODE_ENV !== 'production' &&
+    !!props.css &&
```
We already know that this is truthy here - the first `if` in this function already assumes that and it works
```suggestion
```

mitchellhamilton replied:
pretty sure this is needed, the check at the start of the function checks if it's a string first and then does things. The places where this function is called only check that props has an own prop named `css`

Andarist replied:
ah, ok - makes sense, we indeed allow this `css={null}`

Andarist reviewed (commented):
I left a couple of minor comments/questions. It looks great to me ❤️

srmagura commented:
Thanks! Please see my latest changes as well as my replies to a few of your comments.

Andarist commented on packages/react/src/get-label-from-stack-trace.js:
```diff
@@ -0,0 +1,54 @@
+  identifier.replace(/\$/g, '-')
+
+export const getLabelFromStackTrace = (stackTrace: string): ?string => {
+  // console.log(stackTrace)
```
```suggestion
```
